File: results/dnd-rest-benchmark/lifecycle-runs/20260731T062307Z_claude_claude-sonnet-5_typescript-node/src/routes/campaigns.ts
// Campaigns and their nested characters/events. `hasCampaign` is also used
// by routes/dm.ts, since every DM tool operates within a campaign.
import type { ServerResponse } from "node:http";
import { db } from "../db.js";
import { sendJson } from "../http.js";
import { isPlainObject, isValidInt } from "../validation.js";

interface CampaignRecord {
  id: string;
  name: string;
  dm: string;
}

interface CampaignCharacterRecord {
  campaignId: string;
  id: string;
  name: string;
  level: number;
  class: string;
}

interface CampaignEventRecord {
  campaignId: string;
  id: string;
  kind: string;
  summary: string;
}

export function hasCampaign(id: string): boolean {
  const row = db.prepare("SELECT 1 FROM campaigns WHERE id = ?").get(id);
  return row !== undefined;
}

function getCampaign(id: string): CampaignRecord | undefined {
  const row = db.prepare("SELECT id, name, dm FROM campaigns WHERE id = ?").get(id) as
    | { id: string; name: string; dm: string }
    | undefined;
  if (!row) return undefined;
  return { id: row.id, name: row.name, dm: row.dm };
}

function saveCampaign(campaign: CampaignRecord): void {
  db.prepare("INSERT INTO campaigns (id, name, dm) VALUES (?, ?, ?)").run(campaign.id, campaign.name, campaign.dm);
}

export function hasCampaignCharacter(campaignId: string, id: string): boolean {
  const row = db.prepare("SELECT 1 FROM campaign_characters WHERE campaign_id = ? AND id = ?").get(campaignId, id);
  return row !== undefined;
}

function saveCampaignCharacter(character: CampaignCharacterRecord): void {
  db.prepare(
    "INSERT INTO campaign_characters (campaign_id, id, name, level, class) VALUES (?, ?, ?, ?, ?)",
  ).run(character.campaignId, character.id, character.name, character.level, character.class);
}

function getCampaignCharacters(campaignId: string): CampaignCharacterRecord[] {
  const rows = db
    .prepare("SELECT campaign_id, id, name, level, class FROM campaign_characters WHERE campaign_id = ?")
    .all(campaignId) as { campaign_id: string; id: string; name: string; level: number; class: string }[];
  return rows.map((row) => ({
    campaignId: row.campaign_id,
    id: row.id,
    name: row.name,
    level: row.level,
    class: row.class,
  }));
}

function hasCampaignEvent(campaignId: string, id: string): boolean {
  const row = db.prepare("SELECT 1 FROM campaign_events WHERE campaign_id = ? AND id = ?").get(campaignId, id);
  return row !== undefined;
}

function saveCampaignEvent(event: CampaignEventRecord): void {
  db.prepare("INSERT INTO campaign_events (campaign_id, id, kind, summary) VALUES (?, ?, ?, ?)").run(
    event.campaignId,
    event.id,
    event.kind,
    event.summary,
  );
}

function countCampaignEvents(campaignId: string): number {
  const row = db.prepare("SELECT COUNT(*) AS count FROM campaign_events WHERE campaign_id = ?").get(campaignId) as {
    count: number;
  };
  return row.count;
}

export function handleCreateCampaign(res: ServerResponse, body: unknown): void {
  if (
    !isPlainObject(body) ||
    typeof body.id !== "string" ||
    !body.id ||
    typeof body.name !== "string" ||
    typeof body.dm !== "string"
  ) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  if (hasCampaign(body.id)) {
    sendJson(res, 409, { error: "campaign already exists" });
    return;
  }

  const campaign: CampaignRecord = { id: body.id, name: body.name, dm: body.dm };
  saveCampaign(campaign);

  sendJson(res, 201, { id: campaign.id, name: campaign.name, dm: campaign.dm });
}

export function handleAddCampaignCharacter(res: ServerResponse, campaignId: string, body: unknown): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  if (
    !isPlainObject(body) ||
    typeof body.id !== "string" ||
    !body.id ||
    typeof body.name !== "string" ||
    !isValidInt(body.level, 1, 20) ||
    typeof body.class !== "string"
  ) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  if (hasCampaignCharacter(campaignId, body.id)) {
    sendJson(res, 409, { error: "character already exists" });
    return;
  }

  const character: CampaignCharacterRecord = {
    campaignId,
    id: body.id,
    name: body.name,
    level: body.level,
    class: body.class,
  };
  saveCampaignCharacter(character);

  sendJson(res, 201, {
    id: character.id,
    name: character.name,
    level: character.level,
    class: character.class,
  });
}

export function handleAddCampaignEvent(res: ServerResponse, campaignId: string, body: unknown): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  if (
    !isPlainObject(body) ||
    typeof body.id !== "string" ||
    !body.id ||
    typeof body.kind !== "string" ||
    typeof body.summary !== "string"
  ) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  if (hasCampaignEvent(campaignId, body.id)) {
    sendJson(res, 409, { error: "event already exists" });
    return;
  }

  const event: CampaignEventRecord = {
    campaignId,
    id: body.id,
    kind: body.kind,
    summary: body.summary,
  };
  saveCampaignEvent(event);

  sendJson(res, 201, { id: event.id, kind: event.kind });
}

export function handleGetCampaignState(res: ServerResponse, campaignId: string): void {
  const campaign = getCampaign(campaignId);
  if (!campaign) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  const characters = getCampaignCharacters(campaignId);
  const logCount = countCampaignEvents(campaignId);

  sendJson(res, 200, {
    id: campaign.id,
    name: campaign.name,
    dm: campaign.dm,
    characters: characters.map(({ id, name, level, class: klass }) => ({ id, name, level, class: klass })),
    log_count: logCount,
  });
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T052954Z_claude_claude-sonnet-5_python-flask/dndapp/routes/campaigns.py
"""Campaign creation plus nested characters, events, and aggregate state."""

from flask import Blueprint, jsonify, request

from ..db import get_db
from ..validation import valid_int, valid_nonempty_str

bp = Blueprint("campaigns", __name__)


@bp.post("/v1/campaigns")
def create_campaign():
    data = request.get_json(silent=True) or {}
    campaign_id = data.get("id")
    name = data.get("name")
    dm = data.get("dm")

    if not isinstance(campaign_id, str) or not campaign_id:
        return jsonify(error="invalid id"), 400
    if not valid_nonempty_str(name):
        return jsonify(error="invalid name"), 400
    if not valid_nonempty_str(dm):
        return jsonify(error="invalid dm"), 400

    conn = get_db()
    try:
        existing = conn.execute(
            "SELECT id FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if existing is not None:
            return jsonify(error="campaign already exists"), 409

        conn.execute(
            "INSERT INTO campaigns (id, name, dm) VALUES (?, ?, ?)",
            (campaign_id, name, dm),
        )
        conn.commit()
    finally:
        conn.close()

    return jsonify(id=campaign_id, name=name, dm=dm), 201


@bp.post("/v1/campaigns/<campaign_id>/characters")
def add_campaign_character(campaign_id):
    conn = get_db()
    try:
        campaign = conn.execute(
            "SELECT id FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if campaign is None:
            return jsonify(error="campaign not found"), 404

        data = request.get_json(silent=True) or {}
        char_id = data.get("id")
        name = data.get("name")
        level = data.get("level")
        char_class = data.get("class")

        if not isinstance(char_id, str) or not char_id:
            return jsonify(error="invalid id"), 400
        if not valid_nonempty_str(name):
            return jsonify(error="invalid name"), 400
        if not valid_int(level):
            return jsonify(error="invalid level"), 400
        if not valid_nonempty_str(char_class):
            return jsonify(error="invalid class"), 400

        existing = conn.execute(
            "SELECT id FROM campaign_characters WHERE campaign_id = ? AND id = ?",
            (campaign_id, char_id),
        ).fetchone()
        if existing is not None:
            return jsonify(error="character already exists"), 409

        conn.execute(
            "INSERT INTO campaign_characters (campaign_id, id, name, level, class) "
            "VALUES (?, ?, ?, ?, ?)",
            (campaign_id, char_id, name, level, char_class),
        )
        conn.commit()
    finally:
        conn.close()

    return jsonify(id=char_id, name=name, level=level, **{"class": char_class}), 201


@bp.post("/v1/campaigns/<campaign_id>/events")
def add_campaign_event(campaign_id):
    conn = get_db()
    try:
        campaign = conn.execute(
            "SELECT id FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if campaign is None:
            return jsonify(error="campaign not found"), 404

        data = request.get_json(silent=True) or {}
        event_id = data.get("id")
        kind = data.get("kind")
        summary = data.get("summary")

        if not isinstance(event_id, str) or not event_id:
            return jsonify(error="invalid id"), 400
        if not valid_nonempty_str(kind):
            return jsonify(error="invalid kind"), 400
        if not isinstance(summary, str) or not summary:
            return jsonify(error="invalid summary"), 400

        existing = conn.execute(
            "SELECT id FROM campaign_events WHERE campaign_id = ? AND id = ?",
            (campaign_id, event_id),
        ).fetchone()
        if existing is not None:
            return jsonify(error="event already exists"), 409

        conn.execute(
            "INSERT INTO campaign_events (campaign_id, id, kind, summary) "
            "VALUES (?, ?, ?, ?)",
            (campaign_id, event_id, kind, summary),
        )
        conn.commit()
    finally:
        conn.close()

    return jsonify(id=event_id, kind=kind), 201


@bp.get("/v1/campaigns/<campaign_id>/state")
def get_campaign_state(campaign_id):
    conn = get_db()
    try:
        campaign = conn.execute(
            "SELECT id, name, dm FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if campaign is None:
            return jsonify(error="campaign not found"), 404

        char_rows = conn.execute(
            "SELECT id, name, level, class FROM campaign_characters WHERE campaign_id = ?",
            (campaign_id,),
        ).fetchall()
        log_count = conn.execute(
            "SELECT COUNT(*) AS c FROM campaign_events WHERE campaign_id = ?",
            (campaign_id,),
        ).fetchone()["c"]
    finally:
        conn.close()

    characters = [
        {
            "id": row["id"],
            "name": row["name"],
            "level": row["level"],
            "class": row["class"],
        }
        for row in char_rows
    ]

    return jsonify(
        id=campaign["id"],
        name=campaign["name"],
        dm=campaign["dm"],
        characters=characters,
        log_count=log_count,
    )
